src/ORM/Table.php

```diff
@@ -2249,20 +2253,40 @@ protected function _saveMany(iterable $entities, $options = []): iterable
                     }
                 });
         } catch (Exception $e) {
-            $cleanup($entities);
+            $cleanupOnFailure($entities);
 
             throw $e;
         }
 
         if ($failed !== null) {
-            $cleanup($entities);
+            $cleanupOnFailure($entities);
 
             throw new PersistenceFailedException($failed, ['saveMany']);
         }
 
+        $cleanupOnSuccess = function (EntityInterface $entity) use (&$cleanupOnSuccess) {
+            $entity->clean();
+            $entity->setNew(false);
+
+            foreach (array_keys($entity->toArray()) as $field) {
+                $value = $entity->get($field);
+
+                if ($value instanceof EntityInterface) {
+                    $cleanupOnSuccess($value);
+                } elseif (is_array($value) && current($value) instanceof EntityInterface) {
+                    foreach ($value as $associated) {
+                        $cleanupOnSuccess($associated);
+                    }
+                }
+            }
```

Comment:
Were nested associations not being marked as clean before?

Reply:
They were but as per the issue users don't want the entities to be cleaned before the Model.afterSaveCommit is triggered. Hence I added a new option to prevent `save()` from doing so when called through `saveMany()`. So now after the event is fired we need to iterate though all the entities and clean them.